src/core/lib/transport/metadata.cc

```diff
@@ -549,14 +549,35 @@ grpc_mdelem grpc_mdelem_from_slices(
 }
 
 grpc_mdelem grpc_mdelem_from_grpc_metadata(grpc_metadata* metadata) {
-  bool changed = false;
+  bool key_changed = false;
   grpc_slice key_slice =
-      grpc_slice_maybe_static_intern(metadata->key, &changed);
+      grpc_slice_maybe_static_intern(metadata->key, &key_changed);
+  bool value_changed = false;
+  grpc_slice* unref_slice = nullptr;
   grpc_slice value_slice =
-      grpc_slice_maybe_static_intern(metadata->value, &changed);
-  return grpc_mdelem_create(
-      key_slice, value_slice,
-      changed ? nullptr : reinterpret_cast<grpc_mdelem_data*>(metadata));
+      grpc_slice_maybe_static_intern(metadata->value, &value_changed);
+  // If key or value changed, but the other didn't.... AND the other is a NOP
```

Comment:
Would the need for this be avoided if we just changed `grpc_slice_ref_internal()` to make a copy if the underlying type is not ref-counted?